File: src/main/java/nsu/auth/AuthController.java
package nsu.auth;

import nsu.auth.AuthService;
import nsu.auth.JwtRequest;
import nsu.auth.JwtResponse;
import nsu.auth.RefreshJwtRequest;
import org.springframework.context.annotation.Bean;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Component;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.util.Objects;

@RestController
@RequestMapping("/api/auth")
public class AuthController {
    public AuthController(AuthService authService) {
        this.authService = authService;
    }

    private final AuthService authService;

    @PostMapping("login")
    public ResponseEntity<JwtResponse> login(@RequestBody JwtRequest authRequest) {
        final JwtResponse token = authService.login(authRequest);
        return ResponseEntity.ok(token);
    }

    @PostMapping("token")
    public ResponseEntity<JwtResponse> getNewAccessToken(@RequestBody RefreshJwtRequest request) {
        final JwtResponse token = authService.getAccessToken(Objects.requireNonNull(request.getRefreshToken()));
        return ResponseEntity.ok(token);
    }

    @PostMapping("refresh")
    public ResponseEntity<JwtResponse> getNewRefreshToken(@RequestBody RefreshJwtRequest request) {
        final JwtResponse token = authService.refresh(Objects.requireNonNull(request.getRefreshToken()));
        return ResponseEntity.ok(token);
    }

}

File: src/main/kotlin/nsu/auth/controllers/AuthController.kt
package nsu.auth.controllers

import nsu.auth.entity.User
import nsu.auth.jwt.JwtResponse
import nsu.auth.requests.*
import nsu.auth.service.AuthService
import nsu.auth.service.UserService
import org.springframework.http.ResponseEntity
import org.springframework.security.access.prepost.PreAuthorize
import org.springframework.web.bind.annotation.*
import java.util.*

@RestController
@RequestMapping("/api/v1/auth")
class AuthController(
    private val authService: AuthService,
    private val userService: UserService
) {
    // creates default user with student role
    // everybody can do this
    @PostMapping("register")
    fun registerUser(@RequestBody user: UserRequest): ResponseEntity<*>{
        return try {
            ResponseEntity.ok(userService.addUser(user))
        } catch (e: Exception) {
            ResponseEntity.badRequest().body(e.message)
        }
    }

    // only admin can do this endpoint, have to add roles to the person
    /*
    * "roles" : ["STUDENT", "TEACHER"]
    * */
    @PreAuthorize("hasRole('ADMIN')")
    @PostMapping("register/admin")
    fun registerUserByAdmin(@RequestBody user: User): ResponseEntity<*> {
        return try {
            ResponseEntity.ok(userService.addUser(user))
        } catch (e: Exception) {
            ResponseEntity.badRequest().body(e.message)
        }
    }

    // delete the user
    @PreAuthorize("hasRole('ADMIN')")
    @DeleteMapping("register")
    fun deleteUser(@RequestBody user: UserDeleteRequest): ResponseEntity<*> {
        return try {
            ResponseEntity.ok(userService.deleteUserByLogin(user.login))
        } catch (e: Exception) {
            ResponseEntity.badRequest().body(e.message)
        }
    }

    /*
    * get a new pair of access and refresh token
    * access token is used to send requests
    * refresh - to update the token
    * */

    @CrossOrigin
    @PostMapping("login")
    fun login(@RequestBody authRequest: JwtRequest?): ResponseEntity<*> {
        return try {
           ResponseEntity.ok(authService.login(authRequest!!))
        } catch (e: Exception) {
            ResponseEntity.badRequest().body(e.message)
        }
    }

    /*
    * Update the access token, when sending refresh one in body
    * */
    @PostMapping("token")
    fun getNewAccessToken(@RequestBody request: RefreshJwtRequest): ResponseEntity<JwtResponse> {
        val token = request.refreshToken?.let { authService.getAccessToken(it) }
        return ResponseEntity.ok(token)
    }

    @PostMapping("refresh")
    fun getNewRefreshToken(@RequestBody request: RefreshJwtRequest): ResponseEntity<JwtResponse> {
        val token = request.refreshToken?.let { authService.refresh(it) }
        return ResponseEntity.ok(token)
    }
}
